### PR TITLE
LeanTask (mini task), new methods for tasks

### PR DESCRIPTION
**Pull request by Laxilef**
Inspired by the work of @dbuezas (https://github.com/nrwiersma/ESP8266Scheduler/pull/20), I added LeanTask to the current version of the scheduler and added some features.
This helps to really save RAM. On working existing projects, even 6 tasks completely clogged the heap, after which the esp8266 crashed.
The PR includes updating the documentation. Added a table with LeanTask testing results to the README.
I also added new [examples](https://github.com/Laxilef/ESP8266Scheduler/tree/master/examples). You can review them before merging branches (if you do).

Briefly:
1. Added AbstractTask with new methods: isEnabled, enable, disable, setInterval, getInterval
2. Task and LeanTask inherit AbstractTask
3. Work with the cont library has been moved to Task
4. "tramponline" function moved: static void SchedullerClass::trampoline
5. ::loop() can now be used (although not recommended)
6. Little refactoring
7. Readme updated
8. Version updated to 0.3 :)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/.vscode
```
Comment (nrwiersma) [inline .gitignore:1]: I would prefer these kind of ignores not be in the project, they should be in your global gitignore.

File: README.md [hunk 2]
```diff
@@ -26,69 +28,155 @@ Scheduler.start(&task);
...
+  }
+} print_task;
 ```
 
-**This function handles the ```delay()``` logic. The parent method should be called.**
+</p>
+</details>
 
-# Documentation
+Tasks can run ```yield()``` and ```delay()``` like they normally would. The ```yield()``` function transfer control to the ESP8266, NOT the scheduler. The ```delay()``` function will tell the scheduler that a delay is needed before the next run. If you REALLY need a delay, use ```::delay()```, but this will block the task and the scheduler.
```
Comment (nrwiersma) [inline README.md:108]:
```suggestion
Tasks can run ```yield()``` and ```delay()``` like they normally would. The ```yield()``` function transfers control to the ESP8266, NOT the scheduler. The ```delay()``` function will tell the scheduler that a delay is needed before the next run. If you REALLY need a delay, use ```::delay()```, but this will block the task and the scheduler.
```

File: examples/subfiles_simple/BlinkTask.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  }
+
+  void loop() {
+    state = state == HIGH ? LOW : HIGH;
+    pinMode(2, state);
+  }
+
+private:
+  uint8_t state;
+};
```
Comment (nrwiersma) [inline examples/subfiles_simple/BlinkTask.h:22]:
```suggestion
};

```

File: examples/subfiles_simple/MemTask.h [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+public:
+  MemTask(bool _enabled = false, unsigned long _interval = 0) : LeanTask(_enabled, _interval) {}
+
+protected:
+  void loop() {
+    Serial.print("Free Heap: ");
+    Serial.print(ESP.getFreeHeap());
+    Serial.println(" bytes");
+  }
+};
```
Comment (nrwiersma) [inline examples/subfiles_simple/MemTask.h:13]:
```suggestion
};

```

File: examples/subfiles_simple/PrintTask.h [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+protected:
+  void loop() {
+    Serial.println("Print Loop Start");
+    delay(5000);
+
+    Serial.println("Print Loop End");
+    delay(5000);
+  }
+};
```
Comment (nrwiersma) [inline examples/subfiles_simple/PrintTask.h:15]:
```suggestion
};

```

File: examples/subfiles_simple/subfiles_simple.ino [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+void loop() {
+  static unsigned long ts = 0;
+
+  if ( millis() - ts > 1000 ) {
+    Serial.println("Main loop");
+    ts = millis();
+  }
+}
```
Comment (nrwiersma) [inline examples/subfiles_simple/subfiles_simple.ino:38]:
```suggestion
}

```

File: library.json [hunk 7]
```diff
@@ -9,5 +9,5 @@
   },
   "frameworks": "arduino",
   "platforms": "espressif",
-  "version": "0.2"
+  "version": "0.3"
```
Comment (nrwiersma) [inline library.json:12]: Could you move this to a separate PR please.

File: library.properties [hunk 8]
```diff
@@ -1,5 +1,5 @@
 name=ESP8266Scheduler
-version=0.2
+version=0.3
```
Comment (nrwiersma) [inline library.properties:2]: Same as above

File: src/Scheduler.cpp [hunk 9]
```diff
@@ -1,27 +1,30 @@
...
 
-Task* SchedulerClass::first = NULL;
-Task* SchedulerClass::current = NULL;
+AbstractTask* SchedulerClass::first = NULL;
+AbstractTask* SchedulerClass::current = NULL;
 uint8_t SchedulerClass::nActiveGroupsIdx = 0;
 uint8_t SchedulerClass::nActiveTasks = 0;
 uint8_t SchedulerClass::scheduler_cycle_id = 0;
 uint8_t SchedulerClass::scheduler_run_group_id = 0;
+bool SchedulerClass::mainLoopRunned = 0;
```
Comment (nrwiersma) [inline src/Scheduler.cpp:18]:
```suggestion
bool SchedulerClass::mainLoopRunning = 0;
```

File: src/Scheduler.h [hunk 10]
```diff
@@ -21,24 +18,22 @@ class SchedulerClass {
...
-  static Task* first;
-  static Task* current;
+  static AbstractTask* first;
+  static AbstractTask* current;
 
   static uint8_t nActiveGroupsIdx;
   static uint8_t nActiveTasks;
   static uint8_t scheduler_cycle_id;
   static uint8_t scheduler_run_group_id;
+  static bool mainLoopRunned;
```
Comment (nrwiersma) [inline src/Scheduler.h:36]:
```suggestion
  static bool mainLoopRunning;
```

File: src/Scheduler.h [hunk 11]
```diff
@@ -21,24 +18,22 @@ class SchedulerClass {
...
 };
 
-extern SchedulerClass Scheduler;
-
-#endif
+extern SchedulerClass Scheduler;
```
Comment (nrwiersma) [inline src/Scheduler.h:39]:
```suggestion
extern SchedulerClass Scheduler;

```

**Review by nrwiersma — CHANGES_REQUESTED** [review]
Looks good so far. A couple comments.

**Review by nrwiersma — APPROVED** [review]
LGTM 🎉

**Comment by Laxilef** [comment]
When can we expect a release? :)

**Comment by nrwiersma** [comment]
I am away at the moment and will be back mid next week. Alternatively you can create a PR bumping the version number and I can do the rest on my phone when I get a chance this weekend. Up to you.
